### PR TITLE
Bug 1829994: Index generate dameonless

### PR DESCRIPTION
**Pull request by kevinrizza**
Enable daemonless index add when `--generate` is specified.

This also fixes a bug where, on multiple runs with --generate specified, the database would fail to build.

Doing this with a partial implementation to unblock podman users to be able to generate the database without shelling out to podman and hitting this bug https://github.com/containers/libpod/issues/5234. In the future, this implementation should be fleshed out so that all of the registry and index commands make this distinction between pull and build runners.

### CODE REVIEW

File: cmd/opm/index/add.go [hunk 1]
```diff
@@ -55,10 +55,12 @@ func addIndexAddCmd(parent *cobra.Command) {
 	}
 	indexCmd.Flags().Bool("skip-tls", false, "skip TLS certificate verification for container image registries while pulling bundles")
 	indexCmd.Flags().StringP("binary-image", "i", "", "container image for on-image `opm` command")
-	indexCmd.Flags().StringP("container-tool", "c", "podman", "tool to interact with container images (save, build, etc.). One of: [docker, podman]")
+	indexCmd.Flags().StringP("build-tool", "c", "podman", "tool to build container images (build). One of: [docker, podman]")
+	indexCmd.Flags().StringP("pull-tool", "p", "none", "tool to pull container images (save, pull, etc.). One of: [none, docker, podman]")
 	indexCmd.Flags().StringP("tag", "t", "", "custom tag for container image being built")
 	indexCmd.Flags().Bool("permissive", false, "allow registry load errors")
 	indexCmd.Flags().StringP("mode", "", "replaces", "graph update mode that defines how channel graphs are updated. One of: [replaces, semver, semver-skippatch]")
+	indexCmd.Flags().String("container-tool", "", "container-tool has been deprecated. Please select a build and pull tool separately.")
```
Comment (njhale) [inline cmd/opm/index/add.go:63]: Note for later: cobra has a way to mark options as deprecated:
```go
indexCmd.Flags().StringP("container-tool", "c", "podman", "tool to interact with container images (save, build, etc.). One of: [docker, podman]")
if err := rootCmd.Flags().MarkDeprecated("container-tool", "ignored in favor of separate build-tool and pull-tool options"); err != nil {
		logrus.Panic(err.Error())
}
```

This has some nice properties like preserving the help when referenced, but omitting it from the standard help output.

File: cmd/opm/index/add.go [hunk 2]
```diff
@@ -96,15 +98,29 @@ func runIndexAddCmdFunc(cmd *cobra.Command, args []string) error {
...
+	if err != nil {
+		return err
+	}
+
+	containerTool, err := cmd.Flags().GetString("container-tool")
+	if err != nil {
+		return err
+	}
+
+	if containerTool != "" {
```
Comment (njhale) [inline cmd/opm/index/add.go:120]: Any thoughts on how backwards compatible we want to be? If we explode here we end up breaking existing clients. Though we can do a major version bump after to be dogmatic about semver.

File: cmd/opm/index/add.go (hunk 1 above)
Comment (kevinrizza) [reply cmd/opm/index/add.go:63]: Ah, nice. I'll update this to include it.

File: cmd/opm/index/add.go (hunk 2 above)
Comment (kevinrizza) [reply cmd/opm/index/add.go:120]: I was sort of back and forth on this. I think that's fair, maybe we want to just treat this as an alias and use it to overwrite defaults if the two new flags haven't been specified?

Comment (njhale) [reply cmd/opm/index/add.go:120]: SGTM, but is there a convention for telling the difference between unset and set-to-default?

**Comment by Bowenislandsong** [comment]
/lgtm
/hold
holding for second review

**Comment by kevinrizza** [comment]
@ecordell updated based on your thoughts. ptal.

File: pkg/lib/indexer/indexer.go [hunk 3]
```diff
@@ -339,8 +340,6 @@ func copyDatabaseTo(databaseFile, targetDir string) (string, error) {
 		if err := os.MkdirAll(targetDir, 0777); err != nil {
 			return "", err
 		}
-	} else {
-		return "", err
 	}
```
Comment (ecordell) [inline pkg/lib/indexer/indexer.go:343]: Why is this removed?

File: pkg/lib/indexer/indexer.go [hunk 4]
```diff
@@ -353,7 +352,7 @@ func copyDatabaseTo(databaseFile, targetDir string) (string, error) {
 	dbFile := path.Join(targetDir, defaultDatabaseFile)
 
 	// define the path to copy to the database/index.db file
-	to, err := os.OpenFile(dbFile, os.O_RDWR|os.O_CREATE, 0666)
+	to, err := os.OpenFile(dbFile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0666)
```
Comment (ecordell) [inline pkg/lib/indexer/indexer.go:355]: Doesn't this change mean that if `copy` fails, you'll be left with an empty file instead of your previous index?
